dynamic_programming/matrix_chain_multiplication.py

```diff
@@ -0,0 +1,90 @@
+"""
+Find the minimum number of multiplications needed to multiply chain of matrices.
+Reference: https://www.geeksforgeeks.org/matrix-chain-multiplication-dp-8/
+
+The algorithm has interesting real-world applications. Example:
+1. Image transformations in Computer Graphics as images are composed of matrix.
+2. Solve complex polynomial equations in the field of algebra using least
+processing power.
+3. Calculate overall impact of macroeconomic decisions as economic
+equations involve number of variables.
+4. Self-driving car navigation can be made more accurate as matrix multiplication
+can accurately determine position and orientation of obstacles in short time.
+
+Python doctests can be run with the following command:
+python -m doctest -v matrix_chain_multiply.py
+
+Given a sequence arr[] that represents chain of 2D matrices such that
+the dimension of ith matrix is arr[i-1]*arr[i].
+So suppose arr = [40, 20, 30, 10, 30] means we have 4 matrices of
+dimensions 40*20, 20*30, 30*10 and 10*30.
+
+matrix_chain_multiply() returns an integer denoting
+minimum number of multiplications to multiply the chain.
+
+We do not need to perform actual multiplication here.
+We only need to decide the order in which to perform the multiplication.
+
+Hints:
+1. Number of multiplications (ie cost) to multiply 2 matrices
+of size m*p and p*n is m*p*n.
+2. Cost of matrix multiplication is neither associative ie (M1*M2)*M3 != M1*(M2*M3)
```

Comment:
This is just incorrect. If the individual multiplications are well-defined, then matrix multiplication _is_ associative.